packages/pwa-buildpack/lib/Utilities/__tests__/createFileHash.spec.js

```diff
@@ -0,0 +1,54 @@
+const path = require('path');
+const fs = require('fs');
+
+const createFileHash = require('../createFileHash');
+
+const DEFAULT_SAMPLE_FILE_PATH = path.join(
+    path.resolve(__dirname).split('/packages')[0],
+    'packages',
+    'pwa-buildpack',
+    'lib',
+    'Utilities',
+    '__tests__',
+    'sampleFile.txt'
+);
+
+test('Should return a promise that resolves to a string', async () => {
+    const returnValue = createFileHash(DEFAULT_SAMPLE_FILE_PATH);
+
+    expect(returnValue instanceof Promise).toBeTruthy();
+
+    const hash = await returnValue;
+
+    expect(typeof hash).toBe('string');
+});
+
+test('Should return same hash if the contents of the file has not changed', async () => {
+    const oldValue = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);
+    const newValue = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);
+
+    expect(oldValue === newValue).toBeTruthy();
+});
+
+test('Should return different value if the contents of the same file have changed', async () => {
+    const oldValue = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);
+
+    fs.appendFileSync(`${__dirname}/sampleFile.txt`, ' Venia here.');
+
+    const newValue = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);
+
+    fs.truncateSync(`${__dirname}/sampleFile.txt`, 11);
```

Comment:
We should mock the relevant methods of the `fs` and `crypto` modules instead of actually changing the file on disk here.

Reply:
We are relying on Crypto and FS to create a hash and if we are mocking them, then we are not testing hash creation logic. We need to test if the file has changed, which will often happen while development, the hash changes.

Reply:
I'm bringing it up mostly to highlight unit test best practices. As much as possible we don't want to be making actual network calls or reading from actual files on disk.

Might we be able to simplify and use [digest](https://nodejs.org/api/crypto.html#crypto_hash_digest_encoding) instead?

```
// createFileHash.js
function createFileHash(filePath) {
  const contents = fs.readFileSync(filePath, { encoding: 'utf8' });

  const hash = crypto.createHash('md5');
  hash.update(contents);
  return hash.digest('hex');
}
```

This way in our test we can mock `readFileSync` to just return a string - we trust that the `fs.readFileSync` works correctly, we don't need to test it - and then assert on the result of the function?

```
// createFileHash.spec.js

// mock fs.readFileSync to return 'abc' the first time and 'def' the second time.

const output1 = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);
const output2 = await createFileHash(DEFAULT_SAMPLE_FILE_PATH);

expect(output1).not.toBe(output2);
```

This way we're still testing the "file contents changed" case, but we don't actually have to read from a file on disk :)

Reply:
The problem with readFileSync is first it eats a huge chunk of memory because it tries to load the whole file at a time and the second is it is sync, meaning it will stop the main thread execution till the file is read completely. On the other hand, if we are using streams like in createReadStream the memory footprint is very small.

Sample comparison:

![image](https://user-images.githubusercontent.com/35203638/65538840-d5e92880-decd-11e9-81dd-1236d2114609.png)

On the other hand, mocking the whole stream object to be returned by the mocked createFileHash is too verbose and diminishes the benefit of using this function.

But I get your point, in the case of the 3rd test, I was creating a new file, editing it to check the hash change effect. We can achieve that with mocking without new files. I'll try changing that test if possible.

Reply:
> mocking the whole stream object just to prove a point is not economical

Totally agree, I don't want to do that either.